### ISSUE TITLE
[REQUEST]: always pass an explicit target to zig builds to guarantee consistent assembly

### ISSUE THREAD

**Issue by xdBronch** [compiler-explorer/compiler-explorer#5540]
### Is your feature request related to a problem? Please describe

following ziglang/zig#17357 and closely related to #3936, zig by default compiles for the native target but since compiler explorer may not guarantee that every compilation is done on the same system, this can cause discrepancies in the generated assembly.

### Describe the solution you'd like

zig compilations should specify a `-target` by default. as far as i can tell, the zig CLI simply uses the last invocation of `-target` and similar flags so this shouldnt cause any problems for users wanting to specify their own target

### Describe alternatives you've considered

n/a

### Additional context

https://github.com/ziglang/zig/issues/12458#issuecomment-1217181522

**Comment by andrewrk** [comment]
Hi compiler explorer friends, I think it would be nice to always pass a `-target` flag to Zig if one is not explicitly provided by the user of the website. We're starting to get regular issues filed like this:
* https://github.com/ziglang/zig/issues/17357
* https://github.com/ziglang/zig/issues/17685

**Comment by jeremy-rifkin** [comment]
Does zig optimize for the host architecture by default? CE warns on `-march=native` for other compilers so it would make sense to override this if it's a default, or at least provide a warning about it.

**Comment by partouf** [comment]
A warning would make a lot more sense to me, yeah

**Comment by andrewrk** [comment]
Yes it optimizes for the host architecture by default.

**Comment by jeremy-rifkin** [comment]
I'm not sure what the best design choice is, a default target or a warning, both seem reasonable to me and shouldn't be too hard technically. If we provided a default target, what would you want to default to?

**Comment by partouf** [comment]
Yeah I'm wondering about, would `-target x86_64-linux` also apply `-mcpu generic`? because if not, then `-mcpu generic` would make more sense

**Comment by xdBronch** [comment]
I believe the default CPU is `baseline` when passing an explicit target. I think that would be fine since `generic` doesn't have many extensions so it would generate worse asm than what I think most people would expect.

**Comment by andrewrk** [comment]
> I'm not sure what the best design choice is, a default target or a warning, both seem reasonable to me and shouldn't be too hard technically.

It should be a default target, because you are caching the output, which only makes sense to do to a pure function. You need the default target in order for it to be a pure function, otherwise you have a secret input parameter which is the server that happens to run the compiler process.

> If we provided a default target, what would you want to default to?

`x86_64-linux`, which is equivalent to what all the other compilers on compiler-explorer are defaulting to.

Then you can hit up ARM and tell them you'll change the default to aarch64 if they donate $X/month to compiler-explorer :sunglasses:
